### PR TITLE
Instant Search: Render interface on search pages

### PR DESCRIPTION
**Pull request by jsnmoon**
#### Changes proposed in this Pull Request:

* When a user performs a search using a search widget, the search page should be enhanced with the new instant search interface. Note that this feature is gated by a `JETPACK_SEARCH_PROTOTYPE` define.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Yes, this is part of the Jetpack Instant Search project: p3QzjZ-V3.

#### Testing instructions:
* Add `define( "JETPACK_SEARCH_PROTOTYPE", true );` to your wp-config.php.
* Ensure that your site has the Jetpack Pro plan and has Jetpack Search enabled.
* Enter a query into a search widget. 
Alternatively, navigate to a search page like `/?s=this+is+a+search`.
* Ensure that you see a new Instant Search interface replacing the `main` content area on the left side of the page:

Instant_Search_Test

#### Proposed changelog entry for your changes:
* No changelog entry necessary; these changes will not be accessible to our users.

### CODE REVIEW

**Review by gibrown — DISMISSED** [review]
Tried it out and it is working well. I have lots of UX thoughts, but don't think those should hold up merging this PR. I really like the rebranding to "Instant Search", including changing the paths.

So LGTM.

**Comment by jsnmoon** [comment]
ESLint is throwing errors for Instant Search`s JavaScript files because they're being scored against ES5 linting criteria.

Once #13184 lands, I'll tweak our ESLint configuration to score Instant Search's JavaScript files using ES2015+ linting criteria.

**Comment by jeherve** [comment]
I am afraid we're going to need #13184 so the tests can pass on this PR.

> Enter a query into a search widget.
Alternatively, navigate to a search page like /?s=this+is+a+search.

How do I know the search results page I'm looking at is powered by Jetpack Instant Search?

**Review by jeherve — CHANGES_REQUESTED** [review]
Not directly related to this PR, but it may be something we can fix in here: right now I keep getting the following warning on `master`:

```
[06-Aug-2019 16:47:04 UTC] PHP Warning:  filemtime(): stat failed for _inc/search/dist/jp-search.bundle.js in /var/www/html/wp-content/plugins/jetpack/modules/search/class.jetpack-search.php on line 220
```

Should we replace `filemtime( $file )` by `filemtime( JETPACK__PLUGIN_DIR . $file )` on line 227 here?

**Comment by jsnmoon** [comment]
> Should we replace filemtime( $file ) by filemtime( JETPACK__PLUGIN_DIR . $file ) on line 227 here? (@jeherve)

Agreed; fixed!

> How do I know the search results page I'm looking at is powered by Jetpack Instant Search?

The Instant Search interface will replace the `main` content area on the main section on the left side of the page, like so:

Instant_Search_Test

It should also take a second (or two) for the Instant Search interface to inject onto the page (since it waits for the `DOMContentLoaded` event before doing anything).

**Comment by jeherve** [comment]
> Once #13184 lands, I'll tweak our ESLint configuration to score Instant Search's JavaScript files using ES2015+ linting criteria.

It has now been merged.

**Review by jeherve — DISMISSED** [review]
Since we now have a full directory (`modules/search/instant-search`) that won't be needed in the plugin we ship to WordPress.org, you'd need to add that directory to `.svnignore`.

This will probably need a rebase as well to fix the linting errors that cause the tests to fail.

**Comment by jsnmoon** [comment]
> Since we now have a full directory (modules/search/instant-search) that won't be needed in the plugin we ship to WordPress.org, you'd need to add that directory to .svnignore.
> This will probably need a rebase as well to fix the linting errors that cause the tests to fail.

I added `modules/search/instant-search` to .svnignore and resolved the linting errors with a rebase.

This is ready for another round of 👀!

**Review by gibrown — DISMISSED** [review]
WFM. I'm not sure what I think about keeping the search box in the sidebar yet (mostly because 90% of sites don't have a sidebar). I do agree that on the one theme I tried just typing in the sidebar felt pretty good.

But there is a lot of improvements here, so I think we should get it merged.

**Comment by jsnmoon** [comment]
Screen Shot 2019-08-20 at 5 05 55 PM

Added search results styling that renders the post title, author, publish date, excerpt, and comment count.

File: modules/search/instant-search/components/search-result.jsx
Comment (jeherve) [inline modules/search/instant-search/components/search-result.jsx]: I am not sure I understand why we need the links to open in a new tab here. Since we are on the same site, could we open them as is?

File: modules/search/instant-search/components/search-result.jsx
Comment (jeherve) [inline modules/search/instant-search/components/search-result.jsx]: I seem to have encoding issues here:

image

File: modules/search/instant-search/components/search-result.jsx
Comment (jeherve) [inline modules/search/instant-search/components/search-result.jsx]: Same encoding issues as above:

image

File: modules/search/class.jetpack-search.php
Comment (jeherve) [inline modules/search/class.jetpack-search.php]: I think it would be nice to enqueue an unminified version of the CSS when `SCRIPT_DEBUG` is set to true. We usually do that in Jetpack, to help with debugging:
https://github.com/Automattic/jetpack/blob/b96f66a73398cd79dd460127fa07855ee4b85fcf/_inc/social-logos.php#L10

**Review by jeherve — APPROVED** [review]
This works well for me. I found a few issues, but nothing blocking. I think those can be addressed in a future PR.

********

I noted some differences in the search results, but I can't explain why. This was for blog ID 53262711.

If you look at the results here (not running the new search interface right now):
https://jeremy.hu/?s=marvel&submit=Search&orderby=relevance&order=DESC
You'll see that things start being different from the 3rd result, compared with the results you'd get if you were running a search using this PR:
https://public-api.wordpress.com/rest/v1.3/sites/53262711/search?query=marvel&fields[0]=author&fields[1]=comment_count&fields[2]=date&fields[3]=excerpt_html&fields[4]=gravatar_url&fields[5]=permalink.url.raw&fields[6]=title_html

Do you have any idea why this would happen? Is it because it hits 2 different versions of the endpoint, or is it something else?

********

I use the Jetpack Search widget on my site:

image

The new search interface retains a searchform on the search results page, but that searchform does not include the filters displayed in the widget. I think they should be included.

********

Note that Fusion is going to error out right now, because D29592-code hasn't been committed to WordPress.com so the files are now out of sync. Ideally you'd want to aim to keep them in sync at all times. I would recommend working on merging this diff as well as D30956-code when you can. 

For now, I've forced Fusion to accept this PR.

File: modules/search/instant-search/components/search-result.jsx
Comment (gibrown) [reply modules/search/instant-search/components/search-result.jsx]: Ya i've seen this... i think it has to do with the indexing...

**Comment by gibrown** [comment]
@jeherve ya it uses a different search algorithm and a different index on the new endpoint. So that is why you are seeing different results. Adding support for filtering is getting worked on in this PR: https://github.com/Automattic/jetpack/pull/13195

File: modules/search/instant-search/components/search-result.jsx
Comment (jsnmoon) [reply modules/search/instant-search/components/search-result.jsx]: 👍 I'll make note of this in the master issue.

File: modules/search/instant-search/components/search-result.jsx
Comment (jsnmoon) [reply modules/search/instant-search/components/search-result.jsx]: This is now being tracked in #13291.

File: modules/search/instant-search/components/search-result.jsx
Comment (jsnmoon) [reply modules/search/instant-search/components/search-result.jsx]: This is now being tracked in #13291.

File: modules/search/class.jetpack-search.php
Comment (jsnmoon) [reply modules/search/class.jetpack-search.php]: 👍 Added this to the master issue (#13027).

**Pull request merged**
